applications/CoSimulationApplication/custom_utilities/feti_dynamic_coupling_utilities.cpp

```diff
@@ -0,0 +1,693 @@
+        Vector accel_corrections = prod(rUnitResponse, rLagrangeVec);
+        AddCorrectionToDomain(pDomainModelPart, ACCELERATION, accel_corrections, is_implicit);
+
+        // Apply velocity correction
+        accel_corrections *= (gamma * dt);
+        AddCorrectionToDomain(pDomainModelPart, VELOCITY, accel_corrections, is_implicit);
+
+        if (is_implicit)
+        {
+            // Newmark average acceleration correction
+            // gamma = 0.5
+            // beta = 0.25 = gamma*gamma
+            // deltaAccel = accel_correction
+            // deltaVelocity = 0.5 * dt * accel_correction
+            // deltaDisplacement = dt * dt * gamma * gamma * accel_correction
+
+            accel_corrections *= (gamma * dt);
+            AddCorrectionToDomain(pDomainModelPart, DISPLACEMENT, accel_corrections, is_implicit);
+        }
+        else
+        {
+            // FEM central difference correction:
+            // gamma = 0.5
+            // beta = 0.0
+            // deltaAccel = accel_correction
+            // deltaVelocity = 0.5 * dt * accel_correction
+            // deltaVelocityMiddle = dt * accel_correction
+            // deltaDisplacement = dt * dt * accel_correction
+
+            accel_corrections *= 2.0;
+            AddCorrectionToDomain(pDomainModelPart, MIDDLE_VELOCITY, accel_corrections, is_implicit);
+
+            // Apply displacement correction
+            accel_corrections *= dt;
+            AddCorrectionToDomain(pDomainModelPart, DISPLACEMENT, accel_corrections, is_implicit);
+        }
+
+        KRATOS_CATCH("")
+    }
+
+
+    void FetiDynamicCouplingUtilities::AddCorrectionToDomain(ModelPart* pDomain,
+        const Variable<array_1d<double, 3>>& rVariable, const Vector& rCorrection,
+        const bool IsImplicit)
+    {
+        KRATOS_TRY
+
+        auto domain_nodes = pDomain->NodesArray();
+        const SizeType dim = mpOriginDomain->ElementsBegin()->GetGeometry().WorkingSpaceDimension();
+
+        KRATOS_ERROR_IF_NOT(rCorrection.size() == domain_nodes.size() * dim)
+            << "AddCorrectionToDomain | Correction dof size does not match domain dofs";
+
+        if (IsImplicit)
+        {
+            #pragma omp parallel for
+            for (int i = 0; i < static_cast<int>(domain_nodes.size()); i++)
+            {
+                IndexType equation_id = domain_nodes[i]->GetDof(DISPLACEMENT_X).EquationId();
+                array_1d<double, 3>& r_nodal_quantity = domain_nodes[i]->FastGetSolutionStepValue(rVariable);
+                for (size_t dof_dim = 0; dof_dim < dim; ++dof_dim)
+                {
+                    r_nodal_quantity[dof_dim] += rCorrection[equation_id + dof_dim];
+                }
+            }
+        }
+        else
+        {
+            #pragma omp parallel for
+            for (int i = 0; i < static_cast<int>(domain_nodes.size()); i++)
+            {
+                const double nodal_mass = domain_nodes[i]->GetValue(NODAL_MASS);
```

Comment:
I am quite sure this is a bug!
this is an access by Id and not by index!!!

=> better use the parallel utilities, then this should be easier to avoid ;)

=> search for `block_for_each` in the code ;)

Reply:
Thanks for the concern but I'm pretty sure its not a bug 😅   - the 'id' part of it is resolved by getting the correct correction from the correction vector:
```
IndexType equation_id = domain_nodes[i]->GetValue(EXPLICIT_EQUATION_ID);
array_1d<double, 3>& r_nodal_quantity = domain_nodes[i]->FastGetSolutionStepValue(rVariable);
for (size_t dof_dim = 0; dof_dim < dim; ++dof_dim)
{
             r_nodal_quantity[dof_dim] += rCorrection[equation_id + dof_dim];
}
```
But I did add some more safety by putting a .Has(EXPLICIT_EQUATION_ID) before .GetValue(EXPLICIT_EQUATION_ID) 😄

Reply:
done!